File: ios/App/View/Flash/FlashView.swift
import SwiftUI

/// Study-lamp palette, mirrored from FlashPage.styles.css.ts.
private enum FlashPalette {
  static let paper = Color(red: 0.969, green: 0.965, blue: 0.988)
  static let ink = Color(red: 0.137, green: 0.129, blue: 0.212)
  static let inkSoft = Color(red: 0.431, green: 0.416, blue: 0.525)
  static let line = Color(red: 0.886, green: 0.878, blue: 0.937)
  static let violet = Color(red: 0.357, green: 0.310, blue: 0.780)
  static let violetSoft = Color(red: 0.925, green: 0.918, blue: 0.984)
  static let amber = Color(red: 0.898, green: 0.639, blue: 0.239)
  static let green = Color(red: 0.243, green: 0.420, blue: 0.310)
  static let red = Color(red: 0.627, green: 0.294, blue: 0.235)
}

/// Saved progress: deckId -> card front -> state. Stored as JSON in
/// UserDefaults (the native mirror of the web pack's localStorage key).
private enum FlashStore {
  static let key = "flashbot-progress"

  static func load() -> [String: [String: FlashScheduler.CardState]] {
    guard let data = UserDefaults.standard.data(forKey: key),
      let decoded = try? JSONDecoder().decode(
        [String: [String: FlashScheduler.CardState]].self, from: data)
    else { return [:] }
    return decoded
  }

  static func save(_ progress: [String: [String: FlashScheduler.CardState]]) {
    if let data = try? JSONEncoder().encode(progress) {
      UserDefaults.standard.set(data, forKey: key)
    }
  }
}

struct FlashView: View {
  @State private var openDeckId: String?
  @State private var progress = FlashStore.load()

  var body: some View {
    ZStack {
      FlashPalette.paper.ignoresSafeArea()
      if let deck = FlashContent.decks.first(where: { $0.id == openDeckId }) {
        StudySessionView(
          deck: deck,
          progress: $progress,
          onExit: { openDeckId = nil })
      } else {
        deckList
      }
    }
  }

  private var deckList: some View {
    let today = FlashScheduler.dayIndex(Date())
    return ScrollView {
      VStack(spacing: 14) {
        VStack(spacing: 8) {
          Text(FlashContent.title)
            .font(.system(size: 36, weight: .bold, design: .serif))
            .foregroundStyle(FlashPalette.ink)
          Text(FlashContent.tagline)
            .font(.subheadline)
            .foregroundStyle(FlashPalette.inkSoft)
        }
        .padding(.top, 34)
        .padding(.bottom, 12)

        ForEach(FlashContent.decks) { deck in
          deckRow(deck, today: today)
        }

        Text("Spaced repetition, five boxes, no streak guilt. Built with Repobot.")
          .font(.caption)
          .foregroundStyle(FlashPalette.inkSoft)
          .padding(.top, 24)
      }
      .padding(.horizontal, 22)
      .padding(.bottom, 48)
    }
  }

  private func deckRow(_ deck: FlashContent.Deck, today: Int) -> some View {
    let states = statesForDeck(progress, deck: deck)
    let summary = FlashScheduler.deckProgress(states, today: today)
    let fraction = summary.total == 0 ? 0 : Double(summary.mastered) / Double(summary.total)

    return Button {
      openDeckId = deck.id
    } label: {
      HStack(spacing: 14) {
        Text(deck.emoji)
          .font(.system(size: 28))
          .frame(width: 52, height: 52)
          .background(RoundedRectangle(cornerRadius: 14).fill(FlashPalette.violetSoft))
        VStack(alignment: .leading, spacing: 4) {
          Text(deck.title)
            .font(.system(size: 18, weight: .semibold, design: .serif))
            .foregroundStyle(FlashPalette.ink)
          Text(deck.description)
            .font(.caption)
            .foregroundStyle(FlashPalette.inkSoft)
            .multilineTextAlignment(.leading)
          GeometryReader { geo in
            ZStack(alignment: .leading) {
              Capsule().fill(FlashPalette.line)
              Capsule().fill(FlashPalette.violet)
                .frame(width: geo.size.width * fraction)
            }
          }
          .frame(height: 6)
          .padding(.top, 4)
          Text("\(summary.mastered) of \(summary.total) mastered")
            .font(.caption2)
            .foregroundStyle(FlashPalette.inkSoft)
        }
        Spacer()
        if summary.due > 0 {
          Text("\(summary.due) due")
            .font(.caption2.weight(.bold))
            .foregroundStyle(.white)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(Capsule().fill(FlashPalette.amber))
        } else {
          Text("Done today")
            .font(.caption2.weight(.bold))
            .foregroundStyle(FlashPalette.green)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(Capsule().fill(Color(red: 0.910, green: 0.937, blue: 0.914)))
        }
      }
      .padding(16)
      .background(RoundedRectangle(cornerRadius: 16).fill(Color.white))
      .overlay(RoundedRectangle(cornerRadius: 16).strokeBorder(FlashPalette.line))
    }
    .buttonStyle(.plain)
  }
}

/// Card states for a deck, keyed by card front (stable across content edits).
private func statesForDeck(
  _ progress: [String: [String: FlashScheduler.CardState]], deck: FlashContent.Deck
) -> [FlashScheduler.CardState] {
  let saved = progress[deck.id] ?? [:]
  return deck.cards.map { saved[$0.front] ?? FlashScheduler.newCardState() }
}

private struct StudySessionView: View {
  let deck: FlashContent.Deck
  @Binding var progress: [String: [String: FlashScheduler.CardState]]
  let onExit: () -> Void

  @State private var queue: [Int] = []
  @State private var flipped = false
  @State private var reviewed = 0
  @State private var missed = 0

  private var today: Int { FlashScheduler.dayIndex(Date()) }

  var body: some View {
    VStack(spacing: 0) {
      HStack {
        Button(action: onExit) {
          Label(deck.title, systemImage: "arrow.left")
            .font(.footnote.weight(.semibold))
            .foregroundStyle(FlashPalette.violet)
        }
        .buttonStyle(.plain)
        Spacer()
        Text("\(queue.count) to go")
          .font(.footnote)
          .foregroundStyle(FlashPalette.inkSoft)
      }
      .padding(.top, 22)

      if let currentIndex = queue.first {
        let card = deck.cards[currentIndex]
        Spacer()
        cardView(card)
        Spacer()
        HStack(spacing: 12) {
          gradeButton(
            "Again", enabled: flipped,
            background: Color(red: 0.965, green: 0.906, blue: 0.890),
            foreground: FlashPalette.red
          ) { grade(.again, index: currentIndex) }
          gradeButton(
            "Got it", enabled: flipped,
            background: FlashPalette.violet, foreground: .white
          ) { grade(.good, index: currentIndex) }
        }
        .padding(.bottom, 28)
      } else {
        summaryView
      }
    }
    .padding(.horizontal, 22)
    .onAppear {
      queue = FlashScheduler.dueIndices(statesForDeck(progress, deck: deck), today: today)
    }
  }

  private func cardView(_ card: FlashContent.Flashcard) -> some View {
    VStack(spacing: 12) {
      Text(flipped ? "ANSWER" : "QUESTION")
        .font(.caption2.weight(.bold))
        .kerning(1.6)
        .foregroundStyle(flipped ? FlashPalette.amber : FlashPalette.inkSoft)
      Text(flipped ? card.back : card.front)
        .font(.system(size: 26, weight: .semibold, design: .serif))
        .foregroundStyle(flipped ? FlashPalette.paper : FlashPalette.ink)
        .multilineTextAlignment(.center)
      if flipped, let hint = card.hint {
        Text(hint)
          .font(.footnote.italic())
          .foregroundStyle(Color(red: 0.663, green: 0.647, blue: 0.769))
          .multilineTextAlignment(.center)
      }
      if !flipped {
        Text("Tap to reveal")
          .font(.caption)
          .foregroundStyle(FlashPalette.inkSoft)
      }
    }
    .frame(maxWidth: .infinity)
    .frame(minHeight: 280)
    .padding(24)
    .background(
      RoundedRectangle(cornerRadius: 20)
        .fill(flipped ? FlashPalette.ink : Color.white))
    .overlay(RoundedRectangle(cornerRadius: 20).strokeBorder(FlashPalette.line))
    .rotation3DEffect(.degrees(flipped ? 360 : 0), axis: (x: 1, y: 0, z: 0))
    .animation(.easeInOut(duration: 0.3), value: flipped)
    .onTapGesture { flipped.toggle() }
  }

  private func gradeButton(
    _ label: String, enabled: Bool, background: Color, foreground: Color,
    action: @escaping () -> Void
  ) -> some View {
    Button(action: action) {
      Text(label)
        .font(.body.weight(.bold))
        .foregroundStyle(foreground)
        .frame(maxWidth: .infinity)
        .padding(.vertical, 15)
        .background(RoundedRectangle(cornerRadius: 14).fill(background))
    }
    .buttonStyle(.plain)
    .disabled(!enabled)
    .opacity(enabled ? 1 : 0.4)
  }

  private var summaryView: some View {
    let states = statesForDeck(progress, deck: deck)
    let summary = FlashScheduler.deckProgress(states, today: today)
    return VStack(spacing: 10) {
      Spacer()
      Text(reviewed > 0 ? "🎉" : "🌤️")
        .font(.system(size: 44))
      Text(reviewed > 0 ? "Session complete" : "All caught up")
        .font(.system(size: 24, weight: .bold, design: .serif))
        .foregroundStyle(FlashPalette.ink)
      Text(
        reviewed > 0
          ? "Every due card reviewed. The boxes will bring them back right on time."
          : "Nothing is due in this deck today — come back tomorrow."
      )
      .font(.subheadline)
      .foregroundStyle(FlashPalette.inkSoft)
      .multilineTextAlignment(.center)
      HStack(spacing: 28) {
        statBlock("\(reviewed)", label: "reviews")
        statBlock("\(missed)", label: "misses")
        statBlock("\(summary.mastered)/\(summary.total)", label: "mastered")
      }
      .padding(.top, 14)
      Button(action: onExit) {
        Text("Back to decks")
          .font(.body.weight(.bold))
          .foregroundStyle(.white)
          .frame(maxWidth: .infinity)
          .padding(.vertical, 15)
          .background(RoundedRectangle(cornerRadius: 14).fill(FlashPalette.violet))
      }
      .buttonStyle(.plain)
      .padding(.top, 18)
      Spacer()
    }
  }

  private func statBlock(_ number: String, label: String) -> some View {
    VStack(spacing: 2) {
      Text(number)
        .font(.system(size: 26, weight: .bold, design: .serif))
        .foregroundStyle(FlashPalette.violet)
      Text(label)
        .font(.caption2)
        .foregroundStyle(FlashPalette.inkSoft)
    }
  }

  private func grade(_ value: FlashScheduler.Grade, index: Int) {
    let states = statesForDeck(progress, deck: deck)
    let nextState = FlashScheduler.review(states[index], grade: value, today: today)
    var deckProgress = progress[deck.id] ?? [:]
    deckProgress[deck.cards[index].front] = nextState
    progress[deck.id] = deckProgress
    FlashStore.save(progress)
    reviewed += 1
    if value == .again {
      missed += 1
      // Missed cards come back at the end of this session.
      queue = Array(queue.dropFirst()) + [index]
    } else {
      queue = Array(queue.dropFirst())
    }
    flipped = false
  }
}

#Preview {
  FlashView()
}

File: android/app/src/main/kotlin/com/baseapp/android/view/flash/FlashView.kt
package com.baseapp.android.view.flash

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.util.Date

/** Study-lamp palette, mirrored from the web pack's FlashPage.styles.css.ts. */
private object FlashColors {
    val paper = Color(0xFFF7F6FC)
    val ink = Color(0xFF232136)
    val inkSoft = Color(0xFF6E6A86)
    val line = Color(0xFFE2E0EF)
    val violet = Color(0xFF5B4FC7)
    val violetSoft = Color(0xFFECEAFB)
    val amber = Color(0xFFE5A33D)
    val green = Color(0xFF3E6B4F)
    val greenSoft = Color(0xFFE8EFE9)
    val red = Color(0xFFA04B3C)
    val redSoft = Color(0xFFF6E7E3)
    val hint = Color(0xFFA9A5C4)
}

private const val PREFS_NAME = "flashbot"

/**
 * Saved progress: deckId -> card front -> state, one SharedPreferences
 * entry per card ("deckId::front" -> "box,lastReviewedDay"). The native
 * mirror of the web pack's localStorage key.
 */
private object FlashStore {
    fun prefs(context: Context): SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun load(prefs: SharedPreferences): Map<String, Map<String, FlashScheduler.CardState>> {
        val progress = mutableMapOf<String, MutableMap<String, FlashScheduler.CardState>>()
        for ((key, value) in prefs.all) {
            val keyParts = key.split("::", limit = 2)
            val stateParts = (value as? String)?.split(",")
            if (keyParts.size != 2 || stateParts?.size != 2) continue
            val box = stateParts[0].toIntOrNull() ?: continue
            val lastReviewedDay = stateParts[1].toIntOrNull() ?: continue
            progress.getOrPut(keyParts[0]) { mutableMapOf() }[keyParts[1]] =
                FlashScheduler.CardState(box = box, lastReviewedDay = lastReviewedDay)
        }
        return progress
    }

    fun save(prefs: SharedPreferences, deckId: String, front: String, state: FlashScheduler.CardState) {
        prefs.edit()
            .putString("$deckId::$front", "${state.box},${state.lastReviewedDay}")
            .apply()
    }
}

/** Card states for a deck, keyed by card front (stable across content edits). */
private fun statesForDeck(
    progress: Map<String, Map<String, FlashScheduler.CardState>>,
    deck: FlashContent.Deck,
): List<FlashScheduler.CardState> {
    val saved = progress[deck.id] ?: emptyMap()
    return deck.cards.map { saved[it.front] ?: FlashScheduler.newCardState() }
}

/**
 * Home surface for the `flash` pack — the native twin of the web FlashPage.
 * All scheduling lives in [FlashScheduler] so it stays JVM-testable and in
 * lockstep with the web app. Purely client-side: progress persists to
 * SharedPreferences, no network, no stores.
 */
@Composable
fun FlashView() {
    val context = LocalContext.current
    val prefs = remember { FlashStore.prefs(context) }
    var progress by remember { mutableStateOf(FlashStore.load(prefs)) }
    var openDeckId by remember { mutableStateOf<String?>(null) }

    fun record(deck: FlashContent.Deck, cardIndex: Int, state: FlashScheduler.CardState) {
        val front = deck.cards[cardIndex].front
        FlashStore.save(prefs, deck.id, front, state)
        val deckProgress = (progress[deck.id] ?: emptyMap()) + (front to state)
        progress = progress + (deck.id to deckProgress)
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(FlashColors.paper),
        contentAlignment = Alignment.TopCenter,
    ) {
        val openDeck = FlashContent.decks.firstOrNull { it.id == openDeckId }
        if (openDeck != null) {
            StudySession(
                deck = openDeck,
                progress = progress,
                onGrade = { index, state -> record(openDeck, index, state) },
                onExit = { openDeckId = null },
            )
        } else {
            DeckList(progress = progress, onOpen = { openDeckId = it })
        }
    }
}

@Composable
private fun DeckList(
    progress: Map<String, Map<String, FlashScheduler.CardState>>,
    onOpen: (String) -> Unit,
) {
    val today = FlashScheduler.dayIndex(Date())
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(14.dp),
        modifier = Modifier
            .widthIn(max = 640.dp)
            .verticalScroll(rememberScrollState())
            .statusBarsPadding()
            .navigationBarsPadding()
            .padding(horizontal = 22.dp)
            .padding(top = 34.dp, bottom = 48.dp),
    ) {
        Text(
            text = FlashContent.TITLE,
            color = FlashColors.ink,
            fontSize = 36.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Serif,
        )
        Text(
            text = FlashContent.TAGLINE,
            color = FlashColors.inkSoft,
            fontSize = 14.sp,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(bottom = 12.dp),
        )

        FlashContent.decks.forEach { deck ->
            DeckRow(
                deck = deck,
                summary = FlashScheduler.deckProgress(statesForDeck(progress, deck), today),
                onOpen = { onOpen(deck.id) },
            )
        }

        Text(
            text = "Spaced repetition, five boxes, no streak guilt. Built with Repobot.",
            color = FlashColors.inkSoft,
            fontSize = 12.sp,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = 24.dp),
        )
    }
}

@Composable
private fun DeckRow(
    deck: FlashContent.Deck,
    summary: FlashScheduler.DeckProgress,
    onOpen: () -> Unit,
) {
    val fraction = if (summary.total == 0) 0f else summary.mastered.toFloat() / summary.total
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(Color.White)
            .border(1.dp, FlashColors.line, RoundedCornerShape(16.dp))
            .clickable(onClick = onOpen)
            .padding(16.dp),
    ) {
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .size(52.dp)
                .clip(RoundedCornerShape(14.dp))
                .background(FlashColors.violetSoft),
        ) {
            Text(text = deck.emoji, fontSize = 26.sp)
        }
        Column(
            verticalArrangement = Arrangement.spacedBy(4.dp),
            modifier = Modifier
                .weight(1f)
                .padding(horizontal = 14.dp),
        ) {
            Text(
                text = deck.title,
                color = FlashColors.ink,
                fontSize = 18.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = FontFamily.Serif,
            )
            Text(
                text = deck.description,
                color = FlashColors.inkSoft,
                fontSize = 12.sp,
            )
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 4.dp)
                    .height(6.dp)
                    .clip(CircleShape)
                    .background(FlashColors.line),
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth(fraction)
                        .height(6.dp)
                        .clip(CircleShape)
                        .background(FlashColors.violet),
                )
            }
            Text(
                text = "${summary.mastered} of ${summary.total} mastered",
                color = FlashColors.inkSoft,
                fontSize = 11.sp,
            )
        }
        if (summary.due > 0) {
            Badge(text = "${summary.due} due", background = FlashColors.amber, foreground = Color.White)
        } else {
            Badge(text = "Done today", background = FlashColors.greenSoft, foreground = FlashColors.green)
        }
    }
}

@Composable
private fun Badge(text: String, background: Color, foreground: Color) {
    Text(
        text = text,
        color = foreground,
        fontSize = 11.sp,
        fontWeight = FontWeight.Bold,
        modifier = Modifier
            .clip(CircleShape)
            .background(background)
            .padding(horizontal = 10.dp, vertical = 5.dp),
    )
}

@Composable
private fun StudySession(
    deck: FlashContent.Deck,
    progress: Map<String, Map<String, FlashScheduler.CardState>>,
    onGrade: (index: Int, state: FlashScheduler.CardState) -> Unit,
    onExit: () -> Unit,
) {
    val today = FlashScheduler.dayIndex(Date())
    var queue by remember {
        mutableStateOf(FlashScheduler.dueIndices(statesForDeck(progress, deck), today))
    }
    var flipped by remember { mutableStateOf(false) }
    var reviewed by remember { mutableIntStateOf(0) }
    var missed by remember { mutableIntStateOf(0) }

    fun grade(value: FlashScheduler.Grade) {
        val index = queue.firstOrNull() ?: return
        val state = statesForDeck(progress, deck)[index]
        onGrade(index, FlashScheduler.review(state, value, today))
        reviewed += 1
        queue = if (value == FlashScheduler.Grade.AGAIN) {
            missed += 1
            // Missed cards come back at the end of this session.
            queue.drop(1) + index
        } else {
            queue.drop(1)
        }
        flipped = false
    }

    Column(
        modifier = Modifier
            .widthIn(max = 640.dp)
            .fillMaxSize()
            .statusBarsPadding()
            .navigationBarsPadding()
            .padding(horizontal = 22.dp),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 22.dp),
        ) {
            Text(
                text = "← ${deck.title}",
                color = FlashColors.violet,
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .clip(RoundedCornerShape(8.dp))
                    .clickable(onClick = onExit)
                    .padding(4.dp),
            )
            Spacer(modifier = Modifier.weight(1f))
            Text(
                text = "${queue.size} to go",
                color = FlashColors.inkSoft,
                fontSize = 13.sp,
            )
        }

        val currentIndex = queue.firstOrNull()
        if (currentIndex != null) {
            Spacer(modifier = Modifier.weight(1f))
            StudyCard(
                card = deck.cards[currentIndex],
                flipped = flipped,
                onFlip = { flipped = !flipped },
            )
            Spacer(modifier = Modifier.weight(1f))
            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = 28.dp),
            ) {
                GradeButton(
                    label = "Again",
                    enabled = flipped,
                    background = FlashColors.redSoft,
                    foreground = FlashColors.red,
                    modifier = Modifier.weight(1f),
                ) { grade(FlashScheduler.Grade.AGAIN) }
                GradeButton(
                    label = "Got it",
                    enabled = flipped,
                    background = FlashColors.violet,
                    foreground = Color.White,
                    modifier = Modifier.weight(1f),
                ) { grade(FlashScheduler.Grade.GOOD) }
            }
        } else {
            SessionSummary(
                summary = FlashScheduler.deckProgress(statesForDeck(progress, deck), today),
                reviewed = reviewed,
                missed = missed,
                onExit = onExit,
            )
        }
    }
}

@Composable
private fun StudyCard(card: FlashContent.Flashcard, flipped: Boolean, onFlip: () -> Unit) {
    val flip by animateFloatAsState(
        targetValue = if (flipped) 360f else 0f,
        animationSpec = tween(durationMillis = 300),
        label = "cardFlip",
    )
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp, Alignment.CenterVertically),
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(min = 280.dp)
            .graphicsLayer { rotationX = flip }
            .clip(RoundedCornerShape(20.dp))
            .background(if (flipped) FlashColors.ink else Color.White)
            .border(1.dp, FlashColors.line, RoundedCornerShape(20.dp))
            .clickable(onClick = onFlip)
            .padding(24.dp),
    ) {
        Text(
            text = if (flipped) "ANSWER" else "QUESTION",
            color = if (flipped) FlashColors.amber else FlashColors.inkSoft,
            fontSize = 11.sp,
            fontWeight = FontWeight.Bold,
            letterSpacing = 1.6.sp,
        )
        Text(
            text = if (flipped) card.back else card.front,
            color = if (flipped) FlashColors.paper else FlashColors.ink,
            fontSize = 26.sp,
            fontWeight = FontWeight.SemiBold,
            fontFamily = FontFamily.Serif,
            textAlign = TextAlign.Center,
        )
        if (flipped && card.hint != null) {
            Text(
                text = card.hint,
                color = FlashColors.hint,
                fontSize = 13.sp,
                fontStyle = FontStyle.Italic,
                textAlign = TextAlign.Center,
            )
        }
        if (!flipped) {
            Text(
                text = "Tap to reveal",
                color = FlashColors.inkSoft,
                fontSize = 12.sp,
            )
        }
    }
}

@Composable
private fun GradeButton(
    label: String,
    enabled: Boolean,
    background: Color,
    foreground: Color,
    modifier: Modifier = Modifier,
    onTap: () -> Unit,
) {
    Box(
        contentAlignment = Alignment.Center,
        modifier = modifier
            .alpha(if (enabled) 1f else 0.4f)
            .clip(RoundedCornerShape(14.dp))
            .background(background)
            .clickable(enabled = enabled, onClick = onTap)
            .padding(vertical = 15.dp),
    ) {
        Text(
            text = label,
            color = foreground,
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
        )
    }
}

@Composable
private fun SessionSummary(
    summary: FlashScheduler.DeckProgress,
    reviewed: Int,
    missed: Int,
    onExit: () -> Unit,
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp, Alignment.CenterVertically),
        modifier = Modifier.fillMaxSize(),
    ) {
        Text(text = if (reviewed > 0) "🎉" else "🌤️", fontSize = 44.sp)
        Text(
            text = if (reviewed > 0) "Session complete" else "All caught up",
            color = FlashColors.ink,
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Serif,
        )
        Text(
            text = if (reviewed > 0) {
                "Every due card reviewed. The boxes will bring them back right on time."
            } else {
                "Nothing is due in this deck today — come back tomorrow."
            },
            color = FlashColors.inkSoft,
            fontSize = 14.sp,
            textAlign = TextAlign.Center,
        )
        Row(
            horizontalArrangement = Arrangement.spacedBy(28.dp),
            modifier = Modifier.padding(top = 14.dp),
        ) {
            StatBlock(number = "$reviewed", label = "reviews")
            StatBlock(number = "$missed", label = "misses")
            StatBlock(number = "${summary.mastered}/${summary.total}", label = "mastered")
        }
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 18.dp)
                .clip(RoundedCornerShape(14.dp))
                .background(FlashColors.violet)
                .clickable(onClick = onExit)
                .padding(vertical = 15.dp),
        ) {
            Text(
                text = "Back to decks",
                color = Color.White,
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
            )
        }
    }
}

@Composable
private fun StatBlock(number: String, label: String) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(
            text = number,
            color = FlashColors.violet,
            fontSize = 26.sp,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.Serif,
        )
        Text(
            text = label,
            color = FlashColors.inkSoft,
            fontSize = 11.sp,
        )
    }
}
